app/models/adviser.rb

```diff
@@ -48,12 +57,21 @@ def field_order
 
   private
 
+  def check_for_changes
+    @old_firm_id = firm_id_change.first if firm_id_changed?
+  end
+
   def geocode
     if destroyed?
       firm.geocode
     elsif valid?
       GeocodeAdviserJob.perform_later(self)
     end
+
+    if @old_firm_id.present?
+      IndexFirmJob.perform_later(@old_firm_id)
+      @old_firm_id = nil
+    end
```

Comment:
No race condition, the latter code is to ensure the OLD firm is updated, the GeocodeAdviser job will only reindex the new firm.